docs/rfd.md

```diff
@@ -0,0 +1,478 @@
+}
+```
+
+3. Stopping the job:
+
+```go
+_, err := client.Stop(ctx, jobId)
+```
+
+#### CLI
+
+The CLI is the primary means of interacting with the Job Server. It will have a familiar interface to other CLI tools like `docker` or `kubectl`.
+The following examples demonstrate how the CLI will be used:
+
+```bash
+# starting the job server
+$ sudo jobserver serve --listen-address=127.0.0.1:9090 --rbac=rbac-config.yaml \
+  --cacert=ca.crt --cert=tls.crt --key=tls.key
+
+# to use the client, flags to configure cert and key paths, and the server address, are required.
+# these are omitted in the remaining examples for brevity.
+$ jobserver --address=127.0.0.1:9090 --cacert=ca.crt --cert=tls.crt --key=tls.key ...
+
+# starting a job:
+# for commands that don't require flag args, they can be passed as-is:
+$ jobserver run kubectl logs pod/example
+<job-id> # job id is printed
+
+# for commands that do require flag args, the expected '--' delimiter can be used:
+$ jobserver run -- kubectl logs --follow pod/example
+<job-id>
+
+# to start a job with resource limits:
+$ jobserver run \
+  --cpus=100m \ # 100 milli-cores; can also write e.g. '--cpus=2' for 2 cores (no float values)
+  --memory=1Gi --memory-soft-limit=512Mi \ # the familiar suffixes can be used for memory limits
+  --device-read-bps=8:16=200 \ # format is major:minor=value (bytes per second)
+  --device-write-bps=8:16=200,8:0=50 \ # multiple devices can be specified by separating with commas
+  --device-read-iops=... --device-write-iops=... \ # same format, but units are in iops
+  -- go build -o bin/jobserver ./cmd/jobserver
+
+# stopping a job:
+$ jobserver stop <job-id>
+<job-id>
+
+# getting the status of a job:
+$ jobserver status <job-id>
+state: running
+pid: 1234
+start time: 2006-01-02T15:04:05Z07:00
+
+# getting the output of a job:
+$ jobserver logs [--follow] <job-id>
+<output>
+
+# getting a list of all job ids:
+$ jobserver list
+JOB ID      COMMAND       CREATED          STATUS
+<id-1>      kubectl       2006-01-02...    Running
+<id-2>      go            2006-01-02...    Completed
+```
+
+### Implementation Details
+
+#### Streaming Output
+
+Output streaming will be implemented as follows:
+
+1. When a job is started, we will pipe the combined stdout and stderr of its process to a simple in-memory buffer.
+2. When a client initiates a stream by calling `Output()`, we will write the current contents of the buffer to the stream in 4MB chunks. Then, if the client requested to follow the output, the server will keep the stream open, and any subsequent reads from the process's output will be duplicated to all clients that are following the stream, in addition to the server's own buffer.
+
+#### Process Lifecycle
+
+For jobs that do not specify resource limits, they will be run as a child process of the job server itself without any additional isolation. For jobs that do specify resource limits, they will be run in their own cgroup with all of the requested limits.
```

Comment:
Let's simplify and just run any job in a cgroup with some default limits. You can even simplify the API and remove ability for clients to provide custom limits via API.

Reply:
That works, I can have all jobs be run in cgroups. The defaults could just be the (unlimited) inherited defaults from the parent.